rslib/src/image_occlusion/imagedata.rs

```diff
@@ -0,0 +1,115 @@
+        metadata.data = read(path)?;
+        Ok(metadata)
+    }
+
+    pub fn add_image_occlusion_notes(
+        &mut self,
+        image_path: &str,
+        target_deck: DeckId,
+        occlusions: &str,
+        header: &str,
+        notes: &str,
+        tags: Vec<String>,
+    ) -> Result<OpOutput<()>> {
+        // image file
+        let image_bytes = read(image_path)?;
+        let image_filename = Path::new(&image_path)
+            .file_name()
+            .unwrap()
+            .to_str()
+            .unwrap()
+            .to_string();
+
+        let mgr = MediaManager::new(&self.media_folder, &self.media_db)?;
+        let actual_image_name_after_adding = mgr.add_file(&image_filename, &image_bytes)?;
+
+        let image_tag = format!(
+            "<img id='img' src='{}'></img>",
+            &actual_image_name_after_adding
+        );
+
+        self.transact(Op::ImageOcclusion, |col| {
+            let nt = col.get_or_create_io_notetype()?;
+
+            let mut note = nt.new_note();
+            note.set_field(0, occlusions)?;
+            note.set_field(1, &image_tag)?;
+            note.set_field(2, header)?;
+            note.set_field(3, notes)?;
+            note.tags = tags;
+
+            let last_deck = col.get_last_deck_added_to_for_notetype(note.notetype_id);
+            let ctx = CardGenContext::new(nt.as_ref(), last_deck, col.usn()?);
+            let norm = col.get_config_bool(BoolKey::NormalizeNoteText);
+            col.add_note_inner(&ctx, &mut note, target_deck, norm)?;
+
+            Ok(())
+        })
+    }
+
+    fn get_or_create_io_notetype(&mut self) -> Result<Arc<Notetype>> {
+        let name = self.tr.notetypes_image_occlusion_name().to_string();
+        let nt = match self.get_notetype_by_name(&name)? {
+            Some(nt) => nt,
+            None => {
+                self.add_io_notetype()?;
+                self.get_notetype_by_name(&name).unwrap().unwrap()
+            }
+        };
+        if nt.fields.len() < 4 {
+            Err(AnkiError::TemplateError {
+                info: "IO notetype must have 4+ fields".to_string(),
+            })
+        } else {
+            Ok(nt)
+        }
+    }
+
+    fn add_io_notetype(&mut self) -> Result<()> {
+        let usn = self.usn()?;
+        let tr = &self.tr;
+        let mut nt = Notetype {
+            name: tr.notetypes_image_occlusion_name().into(),
+            config: NotetypeConfig::new_cloze(),
+            ..Default::default()
+        };
+        let occlusion = tr.notetypes_occlusion();
+        nt.add_field(occlusion.as_ref());
+        let image = tr.notetypes_image();
+        nt.add_field(image.as_ref());
+        let header = tr.notetypes_header();
+        nt.add_field(header.as_ref());
+        let back_extra = tr.notetypes_back_extra_field();
+        nt.add_field(back_extra.as_ref());
+
+        let qfmt = format!(
+            "{{{{cloze:{}}}}}\n{{{{{}}}}}\n<script> anki.setupImageCloze() </script>",
```

Comment:
For the script invocation, how about something like:

```html
<div id=err></div>
<script>
try { 
  anki.setupImageCloze(1)
} catch (exc) { 
  document.getElementById("err").innerHTML = `Error loading image occlusion. Is your Anki version up to date?\n\n${exc}`;
}
</script>
```

That way we have a versioned API, and will display something useful for older clients.

Reply:
I will update it.